src/transformers/models/realm/modeling_realm.py

```diff
@@ -0,0 +1,1805 @@
+            reader_correct=reader_correct,
+            block_idx=predicted_block_index,
+            candidate=predicted_candidate,
+            start_pos=predicted_start,
+            end_pos=predicted_end,
+            hidden_states=outputs.hidden_states,
+            attentions=outputs.attentions,
+        )
+
+
+REALM_FOR_OPEN_QA_DOCSTRING = r"""
+    Args:
+        question (`str`):
+            OpenQA Question.
+        answer_ids (`torch.LongTensor` of shape `(num_answers, answer_length)`, *optional*):
+            Answer ids for computing the marginal log-likelihood loss. Indices should be in `[-1, 0, ..., config.vocab_size]` (see `input_ids` docstring) Tokens with indices set to `-1` are ignored (masked),
+            the loss is only computed for the tokens with labels in `[0, ..., config.vocab_size]`
+        return_dict (`bool`, *optional*):
+            Whether or not to return a [`~file_utils.ModelOutput`] instead of a plain tuple.
+"""
+
+
+class RealmSearcher(RealmPreTrainedModel):
+    def __init__(self, config):
+        # TODO(PVP) - this class has to be removed
+        super().__init__(config)
+        self.embedder = RealmEmbedder(config)
+        self.register_buffer(
+            "block_emb",
+            torch.zeros(()).new_empty(
+                size=(config.num_block_records, config.retriever_proj_size),
+                dtype=torch.float32,
+                device=torch.device("cpu"),
+            ),
+        )
+        self.init_weights()
+
+
+@add_start_docstrings(
+    "A wrapper of `RealmSearcher` and `RealmReader` providing end-to-end open domain question answering.",
+    REALM_START_DOCSTRING,
+)
+class RealmForOpenQA(RealmPreTrainedModel):
+    def __init__(self, config, searcher, reader, retriever):
+        super().__init__(config)
+        self.embedder = searcher.embedder
+        self.reader = reader
+        self.block_emb = searcher.block_emb
+        self.retriever = retriever
+
+    #        self.init_weights()
+
+    @property
+    def beam_size(self):
+        if self.training:
+            return self.config.searcher_beam_size
+        return self.config.reader_beam_size
+
+    @classmethod
+    def from_pretrained(cls, searcher_pretrained_name_or_path, reader_pretrained_name_or_path, retriever, **kwargs):
+        """
+        Args:
+            searcher_pretrained_name_or_path (`str`):
+                Searcher pretrained name or path.
+            reader_pretrained_name_or_path (`str`):
+                Reader pretrained name or path.
+
+        """
+        config = kwargs.pop("config", None) or RealmConfig.from_pretrained(searcher_pretrained_name_or_path, **kwargs)
+        searcher = RealmSearcher.from_pretrained(searcher_pretrained_name_or_path, config=config, **kwargs)
+        reader = RealmReader.from_pretrained(reader_pretrained_name_or_path, config=config, **kwargs)
+        return cls(config, searcher, reader, retriever)
+
+    def save_pretrained(self, save_directory):
```

Comment:
this function should also be deleted completely. We should save the whole model only in **one** checkpoint file

Reply:
If we keep the weights in one checkpoint file i.e. the searcher model is removed and integrated into `RealmForOpenQA` model, it would no longer be intuitive to load original tf checkpoints into the model through ` RealmForOpenQA.from_pretrained(..., from_tf=True)` because this method only allows loading one checkpoint at a time, but the `RealmForOpenQA` requires two separate tf checkpoints which are searcher (retriever) and reader (encoder), to make itself work.

What do you think?

Reply:
PS. I just notice that keeping the searcher separated is not very intuitive for tf checkpoints loading as well since the searcher also requires two tf checkpoints which are block_emb and retriever (searcher) weights to work.

Reply:
Hey @qqaatw, 

We have quite a strict policy in `transformers` that all fine-tuned models will always only be in one model checkpoint file. There is no exception to it. One of the main reasons for this is that otherwise one could not save the whole model in a single model repo (since we would need 2 `pytorch_model.bin` files) and this is sadly a no-go.

I've done the final modifications to the structure of `RealmForOpenQA` that allow to load a single checkpoint file. This has a couple of other upsides:
- we can delete the specific `from_pretrained(...)` and `save_pretrained(...)` methods
- We can add the `init_weights()` correctly under the model
- the model has now the exact same structure than RAG which is very helpful for downstream libraries such as https://github.com/deepset-ai/haystack 

Regarding the potential downsides:
- I don't fully agree that is it not intuitive to load everything in one checkpoint file. Note that we can do the exact same thing (one checkpoint) for TF later on. 
- Also we now have three main components when using RAG:
  - RagEmbedder (embeds the questions)
  - RagRetriever (which is now similar to RAG framework independent as there is no need to rely on a framework like PyTorch for an external database and which is responsible for retrieving the context to produce an answer)
  - RagReader (which reads the context and question embeddings and finds the correct index to produce the answer.
=> I think this architecture is quite intuitive and can be understood when compared to the diagram of the official paper: https://arxiv.org/pdf/2002.08909.pdf

What do you think?

Reply:
I see. Considering the function `load_tf_weights_in_realm` will no longer work for `RealmforOpenQA`, should we add a standalone conversion script that converts the original tf checkpoints into single PyTorch checkpoint in case users want to validate our implementation?

Reply:
Exactly, adding a standalone conversion script would be great :-)